test/system/src/test/java/io/pravega/test/system/ReaderCheckpointWithInternalThreadpool.java

```diff
@@ -0,0 +1,246 @@
+        CompletableFuture<Void> result = CompletableFuture.supplyAsync(
+                () -> readEvents(scope, readerId), readerExecutor);
+        Futures.exceptionListener(result,
+                t -> log.error("Error observed while reading events for reader id :{}", readerId, t));
+
+    }
+
+    private <T extends Serializable> void writeEventsToStream(final String scope, final List<T> events) {
+        ClientConfig clientConfig = Utils.buildClientConfig(controllerURI);
+        try (EventStreamClientFactory clientFactory = EventStreamClientFactory.withScope(scope, clientConfig);
+             EventStreamWriter<T> writer = clientFactory.createEventWriter(STREAM,
+                     new JavaSerializer<T>(),
+                     EventWriterConfig.builder().build())) {
+            for (T event : events) {
+                //String routingKey = String.valueOf(event);
+                log.info("Writing message: {} with routing-key: {} to stream {}", event, "", STREAM);
+                writer.writeEvent("", event);
+            }
+            log.info("All events written successfully to stream {}", STREAM);
+        }
+    }
+
+    private <T extends Serializable> Void readEvents(final String scope, final String readerId) {
+
+        final ClientConfig clientConfig = Utils.buildClientConfig(controllerURI);
+
+        try (EventStreamClientFactory clientFactory = EventStreamClientFactory.withScope(scope, clientConfig);
+             EventStreamReader<T> reader = clientFactory.createReader(readerId,
+                     READER_GROUP_NAME,
+                     new JavaSerializer<T>(),
+                     ReaderConfig.builder().build())) {
+            log.info("Reading events from {}/{} with readerId: {}", scope, STREAM, readerId);
+            EventRead<T> event = null;
+            do {
+                try {
+                    event = reader.readNextEvent(READ_TIMEOUT);
+                    //if (event.getEvent() != null) {
+                    log.info("Read event {}", event.getEvent());
+                    //}
+                    if (event.isCheckpoint()) {
+                        log.info("Read a check point event, checkpointName: {}", event.getCheckpointName());
+                    }
+                } catch (ReinitializationRequiredException e) {
+                    log.error("Exception while reading event using readerId: {}", readerId, e);
+                    fail("Reinitialization Exception is not expected");
+                }
+            } while (event.isCheckpoint() && event.getCheckpointName() != CHECKPOINT2);
+            //stop reading if CHECKPOINT_2 is acked by reader.
+            log.info("No more events from {}/{} for readerId: {}", scope, STREAM, readerId);
+        } //reader.close() will automatically invoke ReaderGroup#readerOffline(String, Position)
+        return null;
+    }
+
+    private URI fetchControllerURI() {
+        Service conService = Utils.createPravegaControllerService(null);
```

Comment:
This is creating a controller service. Probably not what is intended. (As the controller was already started on line 100)
This appears to be copied from somewhere, can you make sure to correct this in all the files.

Reply:
Thats right, used existing precondition logic, let me correct that

Reply:
Its actually creating controller service object (not starting new controller service instance) just to fetch URI from service.

Reply:
I am still confused, on line number 101 we are already getting the controller URI we can directly assign that to controllerURI variable and we can use that. 

in that case we don't need to create this method.